### ISSUE TITLE
Do you have pytorch version code?

### ISSUE THREAD

**Issue by KarlFreecss** [saralajew/cbc_networks#3]
I tried to implement CBC Network via Pytorch and after training, the components don't look like number image.
I also ran other version code implemented by others(https://github.com/EmilioOldenziel/torch-cbc), it also doesn't work.

**Comment by LarsHoldijk** [comment]
Hi Karl. First of all, thank you for reaching out to us with your question. It is always nice to hear that people are interested in our work. 

We have not made a pytorch version of CBCs ourselves. However, we are in close contact with the contributors of the pytorch version you have linked to. As far as we are aware, the pytorch package in question should work similar to the Keras version and produce the same results. 

There are however some important considerations for training a CBC that could influence the interpretability of the final components. In section E.2.2 of the appendix of our [paper](http://papers.nips.cc/paper/8546-classification-by-components-probabilistic-modeling-of-reasoning-over-a-set-of-components), you will find details about an ablation study we performed to determine how different settings of a CBC affect the interpretability of its components. In short, some settings have a huge impact on the final interpretability. A feature extractor with swish activation on conv. layers, a euclidean norm constraint on conv. layers and no usage of batch normalization results in interpretable components. 

Perhaps making these changes will improve your results.

**Comment by KarlFreecss** [comment]
Thank you very much! This information is very useful to me!
I will try to use a Euclidean norm constraint on Conv. layers.
Thank you very much again!

**Comment by saralajew** [comment]
@KarlFreecss: Please note a really important thing. If you try to train interpretable components with a Siamese architecture, then you have to control from which type of inputs does the feature extractor receives an update. It is also mentioned in section E.2.2.
To make it clear: Assume f(x) is your feature extractor. We denote by x an input from the *dataset* and by k a *component*. Because the component is *not* sampled from the dataset the gradient regarding the components k should *not* update the weights of the feature extractor f - because the feature extractor should learn to extract useful features from the dataset. Therefore, the feature extractor should *only* receive updates from the dataset samples x. 
However, the CBC specific weights receive updates from both x and k.

I hope that helps. Here are the lines how we've done this by a nasty hack in Keras. We are not happy with that solution as it is really slow. However, as far as we know, there is no other way to solve this in Keras.

cbc_networks/paper/MNIST/negative_reasoning/full_size_CBC_mnist.py

See the class `class Backbone(object)`. It has a routine that copy weights between two backbones.
Here we initialize two backbones:
```   
# initialize the two paths of the Siamese network
backbone = Backbone()
backbone_fix = Backbone(dependent_backbone=backbone)
```
Here we call the two backbones by an input and an component:
```    
# call input data processing path of the Siamese network
data_output = backbone(data_input)

# call component processing path of the Siamese network
add_components = AddComponents(shape=(n_components,) + component_shape,
                                initializer=RandomUniform(0.45, 0.55),
                                constraint=clip,
                                name='add_components_1')
components = add_components(components_input)
components_output = backbone_fix(components)
```
and here is the part where we copy the weight from the input feature extractor to the component feature extractor so that the updated component feature extractor is ignored:
```
copy_weights = callbacks.LambdaCallback(
on_batch_begin=lambda batch, logs: backbone_fix.copy_weights())
```
Maybe it helps. Let us know if you have a running PyTorch example!

**Comment by KarlFreecss** [comment]
Thank you very much!
Every detail is very important!
I utilizy copy.deepcopy() to ignore components' feature extractor's gradient.
I wrote PyTorch code based on https://github.com/EmilioOldenziel/torch-cbc and seem to get the right result.
(My code is in https://github.com/KarlFreecss/torch-cbc, and the image results are in https://github.com/KarlFreecss/torch-cbc/tree/master/char_img)

**Comment by LarsHoldijk** [comment]
Hi Karl,

Good to hear that everything worked out. There are indeed some details that are important for the successful training of a CBC. If you have any more questions feel free to reach out again, either using GitHub or by email (larsholdijk [at] gmail.com). 

Also, if you would be willing to discuss in which context you are looking to use CBCs, I would be very interested in discussing it over email. Perhaps some of our followup work can also be useful for you.
